### ISSUE TITLE
Physics / Animations 

### ISSUE THREAD

**Issue by nipkownix** [nipkownix/re4_tweaks#22]
I'm opening this issue so we can track these better.

Albert mentioned some issues related to animations and physics:

- [ ] Missing facial expressions in-game (Leon's expression when he's hurt or dies should change)
- [ ] Physics not working inside the inventory screen (causes clipping issues with Leon's jacket and the absence of Ada's black ribbon in the inventory)
- [ ] Leon's physics when he's wearing the tactical vest (pl09) are gone (static hair).

About the inventory physics issue, albert mentioned it was a working in the bio4 demo (for the GC i guess), but they disabled it/removed for the final version.

I've been looking at this, and it appears there's a function for each character that controls these physics.
For Leon, it is `cPlLeon::moveCloth`.
Noping it's call makes it look like it does in inventory screen:

![image](https://user-images.githubusercontent.com/33870163/143248382-09789c7a-645f-4d64-8d6d-a111c1276240.png)

It is called by the `cPlayer::move`, which itself is inside `gameMainLoop`.
`gameMainLoop` isn't called while inside the inventory screen, so we get static cloth.
Not sure this would simply be fixed by having ``cPl**::moveCloth`` inside the inventory screen's main loop, but who knows? 

About the tactical vest stuff: I do know the game applies different physics/cloth parameters based on what player model you're using. It is not something store inside the model itself (in this case pl09), but instead some type of table inside the .exe.
That _other dll_ fixed the issue, as far as I can tell.

**Comment by emoose** [comment]
Hmm, had a look at the demo (for reference seems it's the japanese Taikenban demo, the US demos inventory is all 2D without any character render)
Couldn't actually rotate the char to check if cloth was moving or not, but since cloth did move when inventory opened I guess that meant its active.

Took a look at the .DOL for that demo, eventually found `PlClothMoveLeon` at 0x8013E4C4, opened the inventory & set a breakpoint on that in Dolphin, and seemed to be called from 0x80ADC79C.

Unfortunately GC builds use .REL files which separate the code into different files, kinda like Windows DLLs.
Tried searching for files that had matching bytes but no luck with that, some bytes were probably overwritten when the REL was loaded.
Eventually found some bytes that did match though - seems it's being called from Sscrn.rel, at 0x8050EDB8 if loaded  with IDA into 0x80500000 (not sure why lower bits of that addr are so different to what Dolphin got, but comparing code that dolphin shows seems to confirm that's the spot)

After comparing with the debug build Sscrn.rel that we have symbols for, I'm like 80% sure that func is SsPzzlMain::move, which makes sense. Seems the demo just had an extra block of code that'd run `PlClothMoveLeon` & a couple other functions before the rest of it, maybe hooking that func and adding the call back might get it working.

**Comment by emoose** [comment]
> Leon's physics when he's wearing the tactical vest (pl09) are gone (static hair).

Looks like it's the `cmp     byte ptr [eax+4FC9h], 2` checks inside `cPlLeon::initCloth` & `cPlLeon::moveCloth` that are responsible for this, if you nop the jz right after both of them it seems to allow hair physics to work at least.

`pG+0x4FC9` seems to store the ID of the costume being used (gets set by `PlSetCostume`), not sure why they'd exclude the tactical vest specifically here, maybe there's some issue that enabling physics on it causes.

**Comment by nipkownix** [comment]
Hey! Sorry about just vanishing.
I got bit side-tracked trying to improve the user experience a bit:

![image](https://user-images.githubusercontent.com/33870163/143727315-fe6cf3b0-48d9-4ebc-b531-51ea0e9ae398.png)

>E2: also heres my latest 1.1.0 IDB btw, managed to name around ~50% of functions so far 😄

Yoo! Thanks! Your IDB files have been a life saver :p

>The pic above is just the result of running MotionMove & PlClothMoveLeon - if I add in a check to run PlClothSetLeon once then I get a slightly different result:

Very interesting results. 
Maybe if we can find a way to just reposition Leon's jacket so it doesn't clip though his body in this screen, it would be already good enough. Maybe they removed these calls because it made no sense to have those running in a mostly static screen.

>pG+0x4FC9 seems to store the ID of the costume being used (gets set by PlSetCostume), not sure why they'd exclude the tactical vest specifically here, maybe there's some issue that enabling physics on it causes.

Maybe they didn't even noticed the tactical vest wasn't included here, since you only get it right before the end of the game?
Nice find!

>could be that running at 60FPS is making the pain animation happen too quickly for it to be seen

I haven't checked, but it should be happening at 30 fps too.
Also, Ashley's facial expressions do work, for some reason.

>Do you know any easy way to actually check if chara face has changed though? Only way I've found is rotating camera with mouse, but even then it's kinda hard to see...

That's the same problem I've been having. Only albert could have noticed an issue like this lol

This is what he told me when I asked him for an easy way to see the issue:

>Leon opens his mouth and close the eyes when he gets hurt or dies.
It's not that noticeable in motion, but it really looks odd when he dies from an explosion, for example, he lands face up and you can see a standard and stoical expression.

**Comment by emoose** [comment]
Ah just got to try out the new UI, seems to crash for me when I hit F1 though :/ Tried with both 1.1.0 & 1.0.6, is there something I need to copy in for it to work? I'm just using winmm.dll that it built, with no winmm.ini (since WriteSettings should create that?)

Maybe there's a certain point I should be using it?

**Comment by nipkownix** [comment]
Hmm.. weird, it should just work.

Does this build work?

[winmm.zip](https://github.com/nipkownix/re4_tweaks/files/7645727/winmm.zip)

**Comment by emoose** [comment]
Yeah that seems to be fine, just tried making a new build with Release and that works too, seems it's just Debug that crashes for me.

**Comment by nipkownix** [comment]
Ah, I see. I haven't built it in debug mode for a long time. Not sure what could be the problem there though.

**Comment by nipkownix** [comment]
Never ended up implementing this fix, no.
Will do that, and hopefully also an option to allow the mafia costume in cutscenes before the next HD project update.

**Comment by emoose** [comment]
Still not sure why they specifically excluded the vest from having physics applied though, didn't really test with it that much, so could be some issue appears somewhere which made them disable it.

If there is anything like that maybe we could just hook `cPlLeon::moveCloth` and make it check for whatever condition caused an issue instead (holding certain weapon or whatever)

**Comment by emoose** [comment]
If anyone wants to try out the tactical vest fix, here's a build with that included (+ RaiseInventoryAlloc fix):
[dinput8-tacticalVestPhysics.zip](https://github.com/nipkownix/re4_tweaks/files/8055010/dinput8-tacticalVestPhysics.zip)

From the little that I've tested it seems to work fine, but I'd guess there's probably some reason they chose to exclude it though...
 
(btw if you use Raz0rs trainer it seems to overwrite the area we need to patch inside `cPlLeon::initCloth` & `cPlLeon::moveCloth` back to vanilla every frame for some reason, not sure what's going on there, if you want to use that trainer to add tactical vest to inventory make sure to save game & restart after adding it)

**Comment by nipkownix** [comment]
> but I'd guess there's probably some reason they chose to exclude it though...

I vote on it being an oversight, tbh. Can't really think of any possible issue that could occur with this enabled.

Thanks for implementing it. I'll implement the mafia leon on cutscenes thing then.

**Comment by RiasatSalminSami** [comment]
Ada's dress seems to clip through her thighs in the final cutscene where she takes away the sample from Leon.

Was watching Wii gameplay and the cloth doesn't clip through her thighs there.

**Comment by RiasatSalminSami** [comment]
Someone in re4hd site mentioned that Ashley's ballistics physics becomes less prominent in 60 fps mode. Can anyone check and verify this?

**Comment by emoose** [comment]
`cPlAshley::moveBust` & `cSubChar::moveBust` are ballistics related functions AFAIK. On GC these two are pretty much identical, but on PC it looks like `cSubChar::moveBust` added some code that uses pG+0x70 deltaTime float to help calculate something, but `cPlAshley::moveBust` doesn't have that included for some reason, maybe just needs that copied over.

**Comment by emoose** [comment]
Well found an easy way to copy the delta-time stuff to `cPlAshley`, with that it seems physics on her are slowed down to match 30FPS speed now - I'm not sure if this is what people had an issue with though?

Here's a build for anyone that wants to test (this only changes player-controlled Ashley, follower Ashley code is unchanged):
[dinput8-ashleyBustPhysics.zip](https://github.com/nipkownix/re4_tweaks/files/8084790/dinput8-ashleyBustPhysics.zip)

Enable FixAshleyBustPhysics inside F1 menu for it to work, you can toggle it off/on during gameplay if you like.
